File: server/routes/address.ts
import { Router, Request, Response } from 'express'
import { asyncHandler } from '../utils/asyncHandler'

const router = Router()
const LOQATE_API_KEY = process.env.LOQATE_API_KEY

const LOQATE_FIND_URL = 'https://api.addressy.com/Capture/Interactive/Find/v1.00/json3.ws'
const LOQATE_RETRIEVE_URL = 'https://api.addressy.com/Capture/Interactive/Retrieve/v1.00/json3.ws'

router.get(
  '/search',
  asyncHandler(async (req: Request, res: Response) => {
    if (!LOQATE_API_KEY) {
      res.status(503).json({ success: false, error: 'Loqate API key not configured' })
      return
    }

    const { text, country = 'ZA', limit = '10' } = req.query as {
      text?: string
      country?: string
      limit?: string
    }

    if (!text) {
      res.status(400).json({ success: false, error: 'Address text is required' })
      return
    }

    const params = new URLSearchParams({
      Key: LOQATE_API_KEY,
      Text: text,
      Countries: country,
      Limit: limit,
    })

    const response = await fetch(`${LOQATE_FIND_URL}?${params.toString()}`)
    const data = await response.json()
    res.json({ success: true, data })
  })
)

router.get(
  '/retrieve',
  asyncHandler(async (req: Request, res: Response) => {
    if (!LOQATE_API_KEY) {
      res.status(503).json({ success: false, error: 'Loqate API key not configured' })
      return
    }

    const { id } = req.query as { id?: string }
    if (!id) {
      res.status(400).json({ success: false, error: 'Suggestion id is required' })
      return
    }

    const params = new URLSearchParams({ Key: LOQATE_API_KEY, Id: id })
    const response = await fetch(`${LOQATE_RETRIEVE_URL}?${params.toString()}`)
    const data = await response.json()
    res.json({ success: true, data })
  })
)

router.get(
  '/geocode',
  asyncHandler(async (req: Request, res: Response) => {
    const { q } = req.query as { q?: string }
    if (!q) {
      res.status(400).json({ success: false, error: 'Address query is required' })
      return
    }

    const params = new URLSearchParams({
      format: 'json',
      q,
      limit: '1',
      addressdetails: '1',
    })

    const response = await fetch(`https://nominatim.openstreetmap.org/search?${params.toString()}`, {
      headers: { 'User-Agent': 'LegitifyConveyHub/1.0 (dev)' },
    })
    const data = await response.json()
    res.json({ success: true, data })
  })
)

export default router

File: python_server/routers/address.py
import os
import urllib.parse

import httpx
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

router = APIRouter()

LOQATE_API_KEY = os.getenv("LOQATE_API_KEY")
LOQATE_FIND_URL = "https://api.addressy.com/Capture/Interactive/Find/v1.00/json3.ws"
LOQATE_RETRIEVE_URL = "https://api.addressy.com/Capture/Interactive/Retrieve/v1.00/json3.ws"


@router.get("/search")
async def search_address(request: Request):
    if not LOQATE_API_KEY:
        return JSONResponse(
            status_code=503,
            content={"success": False, "error": "Loqate API key not configured"},
        )

    text = request.query_params.get("text")
    country = request.query_params.get("country", "ZA")
    limit = request.query_params.get("limit", "10")

    if not text:
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Address text is required"},
        )

    params = urllib.parse.urlencode({
        "Key": LOQATE_API_KEY,
        "Text": text,
        "Countries": country,
        "Limit": limit,
    })

    async with httpx.AsyncClient() as client:
        response = await client.get(f"{LOQATE_FIND_URL}?{params}")
        data = response.json()

    return {"success": True, "data": data}


@router.get("/retrieve")
async def retrieve_address(request: Request):
    if not LOQATE_API_KEY:
        return JSONResponse(
            status_code=503,
            content={"success": False, "error": "Loqate API key not configured"},
        )

    id = request.query_params.get("id")
    if not id:
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Suggestion id is required"},
        )

    params = urllib.parse.urlencode({"Key": LOQATE_API_KEY, "Id": id})

    async with httpx.AsyncClient() as client:
        response = await client.get(f"{LOQATE_RETRIEVE_URL}?{params}")
        data = response.json()

    return {"success": True, "data": data}


@router.get("/geocode")
async def geocode_address(request: Request):
    q = request.query_params.get("q")
    if not q:
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Address query is required"},
        )

    params = urllib.parse.urlencode({
        "format": "json",
        "q": q,
        "limit": "1",
        "addressdetails": "1",
    })

    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"https://nominatim.openstreetmap.org/search?{params}",
            headers={"User-Agent": "LegitifyConveyHub/1.0 (dev)"},
        )
        data = response.json()

    return {"success": True, "data": data}
